### PR TITLE
[ET-959] Add help banner for Ticket Settings tab.

### PR DESCRIPTION
**Pull request by rafsuntaskin**
[ET-959]

Screenshot:

Screen Shot 2021-03-18 at 9 38 15 PM

[ET-959]: https://theeventscalendar.atlassian.net/browse/ET-959

### CODE REVIEW

File: src/admin-views/settings/help-banner.php [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+	}
+
+	.help-links-wrapper {
+		display: flex;
+	}
+
+	.help-links-wrapper div {
+		min-width: 50%;
+	}
+
```
Comment (juanfra) [inline src/admin-views/settings/help-banner.php:78]: We need to move these to a stylesheet and try to componetize them, and use BEM format for the CSS classes.

for example: `event-tickets__admin-banner`, `event-tickets__admin-banner-help-links` and so on. Please if you need some context about BEM classes, please feel free to ping me.

File: src/admin-views/settings/help-banner.php [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+				foreach ( $etp_resource_links as $link ) {
+					printf( '<li><a href="%s" target="_blank" rel="nofollow">%s</a></li>', esc_attr( $link['href'] ), esc_html( $link['label'] ) );
+				}
+				?>
+			</ul>
+		</div>
+	</div>
+</div>
+<style>
+	#event-tickets-info {
```
Comment (sc0ttkclark) [inline src/admin-views/settings/help-banner.php:54]: Is this an established thing we're doing in regards to embedding styles or should this be in our admin stylesheet?

File: src/admin-views/settings/help-banner.php [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+					if ( ! $etp_enabled ) {
+						printf( ' - <a class="upgrade-link" href="%s" target="_blank" rel="nofollow">%s</a>', esc_attr( 'https://theeventscalendar.com/products/wordpress-event-tickets/' ) , esc_html__( 'Need To Upgrade?', 'event-tickets' ) );;
+					}
+				?>
+			</h3>
+
+			<ul class="kb-links">
+				<?php
+				foreach ( $etp_resource_links as $link ) {
+					printf( '<li><a href="%s" target="_blank" rel="nofollow">%s</a></li>', esc_attr( $link['href'] ), esc_html( $link['label'] ) );
```
Comment (sc0ttkclark) [inline src/admin-views/settings/help-banner.php:46]:
```suggestion
					printf( '<li><a href="%s" target="_blank" rel="nofollow">%s</a></li>', esc_url( $link['href'] ), esc_html( $link['label'] ) );
```

File: src/admin-views/settings/help-banner.php [hunk 4]
```diff
@@ -0,0 +1,82 @@
+<?php
+/**
+ * Settings help banner
```
Comment (juanfra) [inline src/admin-views/settings/help-banner.php:3]: I would use something more related to "Getting started".

File: src/admin-views/settings/help-banner.php [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+	<p class="help-text"><?php esc_html_e( $help_text ); ?></p>
+
+	<div class="help-links-wrapper">
+		<div class="et-help-links">
+			<h3><?php esc_html_e( 'Beginner Resources', 'event-tickets' )  ?> </h3>
+
+			<ul class="kb-links">
+				<?php
+				foreach ( $et_resource_links as $link ) {
+					printf( '<li><a href="%s" target="_blank" rel="nofollow">%s</a></li>', esc_attr( $link['href'] ), esc_html( $link['label'] ) );
```
Comment (sc0ttkclark) [inline src/admin-views/settings/help-banner.php:28]:
```suggestion
					printf( '<li><a href="%s" target="_blank" rel="nofollow">%s</a></li>', esc_url( $link['href'] ), esc_html( $link['label'] ) );
```

Comment (sc0ttkclark) [reply src/admin-views/settings/help-banner.php:28]: Also need the noop on the links right?

File: src/admin-views/settings/help-banner.php (hunk 3 above)
Comment (sc0ttkclark) [reply src/admin-views/settings/help-banner.php:46]: Also need the noop on the links right?

**Review by juanfra — CHANGES_REQUESTED** [review]
Looking great! I have a couple comments that would be great if we can address. And also to add a readme entry.

File: src/functions/views/provider.php [hunk 6]
```diff
@@ -17,3 +17,70 @@ function tribe_events_tickets_views_v2_is_enabled() {
...
+ *
+ * @since TBD
+ *
+ * @return string
+ */
+function event_tickets_render_settings_banner() {
+
+	$et_resource_links = [
+		[
+			'label' => esc_html__( 'Getting Started Guide', 'event-tickets' ),
```
Comment (sc0ttkclark) [inline src/functions/views/provider.php:34]: Don't need esc_html__ here, we esc_html later

File: src/functions/views/provider.php [hunk 7]
```diff
@@ -17,3 +17,70 @@ function tribe_events_tickets_views_v2_is_enabled() {
 
 	return tribe_events_views_v2_is_enabled();
 }
+
+add_action( 'tribe_settings_before_content_tab_event-tickets', 'event_tickets_render_settings_banner' );
+
+/**
+ * Render the Help banner for the Ticket Settings Tab
```
Comment (sc0ttkclark) [inline src/functions/views/provider.php:24]:
```suggestion
 * Render the Help banner for the Ticket Settings Tab.
```

File: src/admin-views/settings/help-banner.php [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+ * @since TBD
+ *
+ * @var bool $etp_enabled         Event Tickets Plus enabled or not.
+ * @var array $et_resource_links  Knowledgebase links for Event Tickets.
+ * @var array $etp_resource_links Knowledgebase links for Event Tickets.
+ */
+
+$help_text = $etp_enabled
+	? __( 'Thank you for using Event Tickets and Event Tickets Plus! We recommend looking through the settings below so that you can fine tune your specific ticketing needs. Here are some resources that can help.', 'event-tickets' )
+	: __( 'Thank you for using Event Tickets! We recommend looking through the settings below so that you can fine tune your specific ticketing needs. Here are some resources that can help.', 'event-tickets' );
```
[the remark this answers is not in the document]
Comment (sc0ttkclark) [reply src/admin-views/settings/help-banner.php:14]: It escapes later, so this isn't needed

File: src/admin-views/settings/help-banner.php [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+?>
+<div id="event-tickets-info">
+	<h3><?php esc_html_e( 'Getting Started With Tickets', 'event-tickets' ); ?></h3>
+	<p class="help-text"><?php esc_html_e( $help_text ); ?></p>
```
Comment (sc0ttkclark) [inline src/admin-views/settings/help-banner.php:19]:
```suggestion
	<p class="help-text"><?php echo esc_html( $help_text ); ?></p>
```

File: src/functions/views/provider.php [hunk 10]
```diff
@@ -17,3 +17,70 @@ function tribe_events_tickets_views_v2_is_enabled() {
...
+	$context = [
+		'etp_enabled'        => class_exists( 'Tribe__Tickets_Plus__Main' ),
+		'et_resource_links'  => $et_resource_links,
+		'etp_resource_links' => $etp_resource_links,
+	];
+
+	/** @var Tribe__Tickets__Admin__Views $admin_views */
+	$admin_views = tribe( 'tickets.admin.views' );
+
+	return $admin_views->template( 'settings/help-banner', $context );
```
[the remark this answers is not in the document]
Comment (sc0ttkclark) [reply src/functions/views/provider.php:85]: +1 on that

**Review by sc0ttkclark — COMMENTED** [review]
Added some more direction notes here

File: src/admin-views/settings/help-banner.php (hunk 2 above)
Comment (rafsuntaskin) [reply src/admin-views/settings/help-banner.php:54]: We don't have any admin stylesheet loading within the Settings pages. We need to create a new stylesheet and move the styles.

File: src/admin-views/settings/help-banner.php (hunk 5 above)
Comment (paulmskim) [reply src/admin-views/settings/help-banner.php:28]: add `noopener` and `noreferrer` to `rel` if target is `_blank`. This goes for all links in this PR that have a target of `_blank`

File: src/admin-views/settings/help-banner.php [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+				}
+				?>
+			</ul>
+		</div>
+		<div class="etp-help-links">
+			<h3>
+				<?php esc_html_e( 'Advanced Plus Features', 'event-tickets' ); ?>
+				<?php
+					if ( ! $etp_enabled ) {
+						printf( ' - <a class="upgrade-link" href="%s" target="_blank" rel="nofollow">%s</a>', esc_attr( 'https://theeventscalendar.com/products/wordpress-event-tickets/' ) , esc_html__( 'Need To Upgrade?', 'event-tickets' ) );;
```
Comment (paulmskim) [inline src/admin-views/settings/help-banner.php:38]:
```suggestion
						printf( ' - <a class="upgrade-link" href="%s" target="_blank" rel="nofollow">%s</a>', esc_attr( 'https://theeventscalendar.com/products/wordpress-event-tickets/' ), esc_html__( 'Need To Upgrade?', 'event-tickets' ) );
```

File: src/Tribe/Assets.php [hunk 12]
```diff
@@ -325,6 +341,30 @@ public function should_enqueue_admin() {
...
+	 * @since TBD
+	 *
+	 * @return bool
+	 */
+	public function should_enqueue_admin_settings_assets() {
+
+		$admin_helpers = Tribe__Admin__Helpers::instance();
+
+		// Load specifically on Ticket Settings page only.
+		$should_enqueue = $admin_helpers->is_screen() && 'event-tickets' === tribe_get_request_var( 'tab' );
```
Comment (sc0ttkclark) [inline src/Tribe/Assets.php:356]:
```suggestion
		$admin_helpers = Tribe__Admin__Helpers::instance();

		// The list of admin tabs that the plugin hooks into.
		$admin_tabs = [
			'event-tickets',
			'event-tickets-commerce',
		];

		// Load specifically on Ticket Settings page only.
		$should_enqueue = $admin_helpers->is_screen() && in_array( tribe_get_request_var( 'tab' ), $admin_tabs, true );
```

File: src/admin-views/settings/getting-started.php [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+	: __( 'Thank you for using Event Tickets! We recommend looking through the settings below so that you can fine tune your specific ticketing needs. Here are some resources that can help.', 'event-tickets' );
+
+?>
+<div id="event-tickets__admin-banner">
+	<h3><?php echo esc_html__( 'Getting Started With Tickets', 'event-tickets' ); ?></h3>
+	<p class="event-tickets__admin-banner-help-text"><?php echo esc_html__( $help_text ); ?></p>
+
+	<div class="event-tickets__admin-banner-help-links">
+		<div class="event-tickets__admin-banner-et-links">
+			<h3><?php echo esc_html__( 'Beginner Resources', 'event-tickets' )  ?> </h3>
```
Comment (sc0ttkclark) [inline src/admin-views/settings/getting-started.php:23]:
```suggestion
			<h3><?php esc_html_e( 'Beginner Resources', 'event-tickets' ); ?> </h3>
```

File: src/admin-views/settings/getting-started.php [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+				<?php
+				foreach ( $et_resource_links as $link ) {
+					printf( '<li><a href="%s" target="_blank" rel="noopener noreferrer">%s</a></li>', esc_url( $link['href'] ), esc_html( $link['label'] ) );
+				}
+				?>
+			</ul>
+		</div>
+		<div class="event-tickets__admin-banner-etp-links">
+			<h3>
+				<?php echo esc_html__( 'Advanced Plus Features', 'event-tickets' ); ?>
```
Comment (sc0ttkclark) [inline src/admin-views/settings/getting-started.php:35]:
```suggestion
				<?php esc_html_e( 'Advanced Plus Features', 'event-tickets' ); ?>
```

File: src/functions/views/provider.php [hunk 15]
```diff
@@ -17,3 +17,70 @@ function tribe_events_tickets_views_v2_is_enabled() {
...
+
+/**
+ * Render the Help banner for the Ticket Settings Tab.
+ *
+ * @since TBD
+ *
+ * @return string
+ */
+function event_tickets_render_settings_banner() {
+
```
Comment (sc0ttkclark) [inline src/functions/views/provider.php:31]:
```suggestion
```

File: src/functions/views/provider.php [hunk 16]
```diff
@@ -17,3 +17,70 @@ function tribe_events_tickets_views_v2_is_enabled() {
...
 }
+
+add_action( 'tribe_settings_before_content_tab_event-tickets', 'event_tickets_render_settings_banner' );
+
+/**
+ * Render the Help banner for the Ticket Settings Tab.
+ *
+ * @since TBD
+ *
+ * @return string
```
Comment (sc0ttkclark) [inline src/functions/views/provider.php:28]:
```suggestion
 * @return string The help banner HTML content.
```

**Review by sc0ttkclark — CHANGES_REQUESTED** [review]
The code looks good and is ready, but it's missing something the ticket is calling for -- @rafsuntaskin This always renders on the screen, but does not expand/collapse

File: src/admin-views/settings/getting-started.php [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+ * @var array $et_resource_links  Knowledgebase links for Event Tickets.
+ * @var array $etp_resource_links Knowledgebase links for Event Tickets.
+ */
+
+$help_text = $etp_enabled
+	? __( 'Thank you for using Event Tickets and Event Tickets Plus! We recommend looking through the settings below so that you can fine tune your specific ticketing needs. Here are some resources that can help.', 'event-tickets' )
+	: __( 'Thank you for using Event Tickets! We recommend looking through the settings below so that you can fine tune your specific ticketing needs. Here are some resources that can help.', 'event-tickets' );
+
+?>
+<div id="event-tickets__admin-banner">
```
Comment (juanfra) [inline src/admin-views/settings/getting-started.php:17]: Is there any specific reason why you're using and ID here? I'd use class directly.

File: src/resources/postcss/tickets-admin-settings.pcss [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+	a {
+		text-decoration: none;
+	}
+
+	p.event-tickets__admin-banner-help-text {
+		max-width: 80%;
+	}
+
+
+	.event-tickets__admin-banner-help-links {
```
Comment (juanfra) [inline src/resources/postcss/tickets-admin-settings.pcss:19]: Let's move this class out of `event-tickets__admin-banner`

File: src/resources/postcss/tickets-admin-settings.pcss [hunk 19]
```diff
@@ -0,0 +1,31 @@
+
+#event-tickets__admin-banner {
```
Comment (juanfra) [inline src/resources/postcss/tickets-admin-settings.pcss:2]: Let's have this as a class.

File: src/resources/postcss/tickets-admin-settings.pcss [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+	.event-tickets__admin-banner-help-links {
+		display: flex;
+
+		div {
+			min-width: 50%;
+		}
+	}
+
+	.event-tickets__admin-banner-et-links h3,
+	.event-tickets__admin-banner-etp-links h3 {
```
Comment (juanfra) [inline src/resources/postcss/tickets-admin-settings.pcss:28]: Let's have a generic class and use it in both places. Let's move this class out of `event-tickets__admin-banner`

**Review by juanfra — COMMENTED** [review]
Looking awesome, a few more adjustments and we should be good to go 🤘

File: src/admin-views/settings/getting-started.php (hunk 17 above)
Comment (rafsuntaskin) [reply src/admin-views/settings/getting-started.php:17]: No specific reason, moving out as a class.

File: src/resources/postcss/tickets-admin-settings.pcss [hunk 21]
```diff
@@ -14,18 +14,16 @@
 	p.event-tickets__admin-banner-help-text {
 		max-width: 80%;
 	}
+}
 
+.event-tickets__admin-banner-help-links-wrapper {
```
Comment (sc0ttkclark) [inline src/resources/postcss/tickets-admin-settings.pcss:19]: Should this be within the `.event-tickets__admin-banner` for structure or does that not matter?

File: src/resources/postcss/tickets-admin-settings.pcss [hunk 22]
```diff
@@ -14,18 +14,16 @@
...
-
-	.event-tickets__admin-banner-et-links h3,
-	.event-tickets__admin-banner-etp-links h3 {
-		font-size: 14px;
+	div {
+		min-width: 50%;
 	}
 }
+
+.event-tickets__admin-banner-links h3 {
```
Comment (sc0ttkclark) [inline src/resources/postcss/tickets-admin-settings.pcss:27]: Should this be within the `.event-tickets__admin-banner` for structure or does that not matter?

File: src/Tribe/Assets.php [hunk 23]
```diff
@@ -325,6 +341,36 @@ public function should_enqueue_admin() {
 		return ! empty( $post ) && ! empty( $modules ) && in_array( $post->post_type, tribe( 'tickets.main' )->post_types(), true );
 	}
 
+	/**
+	 * Check if we should add the Admin Settings Assets into a Page
```
Comment (sc0ttkclark) [inline src/Tribe/Assets.php:345]:
```suggestion
	 * Check if we should add the Admin Settings Assets onto an admin page.
```

File: src/Tribe/Assets.php [hunk 24]
```diff
@@ -325,6 +341,36 @@ public function should_enqueue_admin() {
...
+		$admin_tabs = [
+			'event-tickets',
+			'event-tickets-commerce',
+		];
+
+		// Load specifically on Ticket Settings page only.
+		$should_enqueue = $admin_helpers->is_screen() && in_array( tribe_get_request_var( 'tab' ), $admin_tabs, true );
+
+		/**
+		 * Allow filtering of where the base Admin Settings Assets will be loaded
```
Comment (sc0ttkclark) [inline src/Tribe/Assets.php:365]:
```suggestion
		 * Allow filtering of whether the base Admin Settings Assets should be loaded.
```

**Review by sc0ttkclark — APPROVED** [review]

**Comment by sc0ttkclark** [comment]
Maybe a postcss enhancement, but optional IMO

File: src/resources/postcss/tickets-admin-settings.pcss (hunk 21 above)
Comment (rafsuntaskin) [reply src/resources/postcss/tickets-admin-settings.pcss:19]: I moved them out as @juanfra suggested in his CR. Not much sure how BEM in PostCSS works.

Comment (juanfra) [reply src/resources/postcss/tickets-admin-settings.pcss:19]: @sc0ttkclark @rafsuntaskin Ideally we should have them outside of it. We're following the same rules we have for the rest of the CSS using BEM, where each block has to be defined independently (with the exception of the wrappers we have, like `tribe-common`). There's more info about this here: https://github.com/the-events-calendar/the-events-calendar/tree/master/src/resources/postcss

And you can check this stylesheet as a reference: https://github.com/the-events-calendar/the-events-calendar/blob/master/src/resources/postcss/views/full/month/_calendar-event.pcss

File: src/functions/views/provider.php [hunk 25]
```diff
@@ -17,3 +17,69 @@ function tribe_events_tickets_views_v2_is_enabled() {
...
+		'etp_enabled'        => class_exists( 'Tribe__Tickets_Plus__Main' ),
+		'et_resource_links'  => $et_resource_links,
+		'etp_resource_links' => $etp_resource_links,
+	];
+
+	/** @var Tribe__Tickets__Admin__Views $admin_views */
+	$admin_views = tribe( 'tickets.admin.views' );
+
+	return $admin_views->template( 'settings/getting-started', $context );
+}
```
Comment (bordoni) [inline src/functions/views/provider.php:85]: This doesn't seem like the correct place for this set of function/hook. I would love this to be in a class.

**Review by sc0ttkclark — APPROVED** [review]

**Review by bordoni — CHANGES_REQUESTED** [review]

**Comment by rafsuntaskin** [comment]
The update looks good to me. Having its own service provider is definitely a Plus 💯

**Review by sc0ttkclark — APPROVED** [review]

File: src/Tribe/Admin/Settings/Service_Provider.php [hunk 26]
```diff
@@ -0,0 +1,102 @@
...
+ * @since   TBD
+ */
+class Service_Provider extends tad_DI52_ServiceProvider {
+	/**
+	 * Register the provider singletons.
+	 *
+	 * @since TBD
+	 */
+	public function register() {
+		$this->container->singleton( 'tickets.admin.manager', self::class );
```
Comment (sc0ttkclark) [inline src/Tribe/Admin/Settings/Service_Provider.php:21]:
```suggestion
		$this->container->singleton( 'tickets.admin.settings', self::class );
```

**Review by sc0ttkclark — APPROVED** [review]

**Pull request merged**